RSDK-794 Client side helper to replace `stub_wrapper`

acmorrow opened the pull request:
Sorry this took a little longer than expected, I ended up trying several different designs before settling on this one. This is the client side analogue of #178.

@stuqdog and @benjirewis - For main review
CC @zaporter-work and @abe-winter for interest.

Thanks to @abe-winter for starting with `stub_wrapper` which pointed to deriving most of the needed info from the pointer-to-member on the stub type.

acmorrow commented on src/viam/sdk/common/client_helper.hpp:
```diff
@@ -0,0 +1,82 @@
+        }
+
+        std::forward<ErrorHandlerCallable>(ehc)(result);
+        std::abort();
```
The `ErrorHandlerCallable` is supposed to throw. But, it might not!

In that case we could either terminate, as I'm doing now, or we could call `default_ehc_` for you. I chose the former because the latter felt like papering over a programmer error.

Definitely interested in others thoughts here.

benjirewis replied:
What if an invocation of the `ClientHelper` had an `ehc` that swallowed an error and logged it or something? Should we just return some nullish value here instead of aborting? To be clear, I agree that we shouldn't call `default_ehc_` in the absence of a throw, but I'm not convinced we should `abort` either.

acmorrow replied:
Unfortunately, I don't think that will work out. If we change this code to `return std::forward<ErrorHandlerCallable(ehc)>(result)`, then *every* ehc will need to have the same return type as `rhc` or this function won't compile. Which would mean we couldn't have a default for `ehc` since every `invoke` handler has its own return type.

So, we can't return, so we must throw, and the `ehc` didn't do that, so either we throw or we abort.

benjirewis replied:
Ah gotcha. I was thinking about something like

```
std::forward<ErrorHandlerCallable>(ehc)(result);
return ["empty"-value-of-whatever-rhc-returns];
```

but yeah seems like that's not possible given there's no template type for the rhc return value. I can just recall MongoDB Go driver methods that would swallow certain errors from the server, and it seems like using `ClientHelper` means if a non-ok gRPC status is returned, some error _must_ be thrown or the program will abort. If we _really_ wanted to swallow an error though, we could just not use `ClientHelper`.

stuqdog replied:
Given that we can't reasonably return anything here, I agree that `abort` makes the most sense. I wonder if there's some value in logging the error message ourselves, just to provide more info to the user?

acmorrow replied:
Sure, that's easy enough.

acmorrow commented on src/viam/sdk/common/client_helper.hpp:
```diff
@@ -0,0 +1,82 @@
+    auto invoke(ResponseHandlerCallable&& rhc, ErrorHandlerCallable&& ehc) {
+        *request_.mutable_name() = client_->name();
+        ::grpc::ClientContext ctx;
+        set_client_ctx_authority(ctx);
```
Note that I don't think this obviates the need for #179 of some form. This class shouldn't be responsible for properly configuring a client context. But it does mean that if we convert the codebase over to `ClientHelper`, there will be fewer places where the factory or wrapper class needs to be injected.

benjirewis replied:
> This class shouldn't be responsible for properly configuring a client context.

If we do convert the whole codebase to `ClientHelper` (including the complex module example), why would we need to configure a client context anywhere else?

benjirewis replied:
Cool that makes sense; thanks!

acmorrow commented on src/viam/sdk/common/client_helper.hpp:
```diff
@@ -0,0 +1,82 @@
+
+template <typename ClientType, typename StubType, typename RequestType, typename ResponseType>
+auto make_client_helper(ClientType* client,
+                        StubType& stub,
```
The choice of reference over pointer here is because the Stub is usually held by `unique_ptr`, and it felt better to write `*stub_` than `stub.get()` everywhere. But it could be easily changed if that's too quirky.

benjirewis replied:
No strong opinion; I don't view `*stub_` as any less strange than `stub.get()`, but references seem more ergonomic than pointers in my cursory C++ knowledge.

stuqdog replied:
seems reasonable to me

acmorrow commented on src/viam/sdk/components/motor/client.cpp:
```diff
@@ -141,66 +81,27 @@ grpc::StatusCode MotorClient::stop(const AttributeMap& extra) {
-        throw std::runtime_error(status.error_message());
-    }
-    return from_proto(response);
+    return make_client_helper(this, *stub_, &StubType::IsPowered)
```
⬆️  - I did not convert `MotorClient::stop` because the behavior is non-standard. As written the `ClientHelper` cannot deal with it, because it splits the status into the good case where you handle a response and a bad case where you handle an error. It may actually be possible to SFINAE in the ability to have the invoke handler take a callback that takes a `grpc::Status` object, which would let us use the helper here.

benjirewis replied:
Yeah I would rather change the `stop` return type than SFINAE our way around it if possible.

acmorrow commented on src/viam/sdk/components/motor/client.cpp:
```diff
@@ -141,66 +81,27 @@ grpc::StatusCode MotorClient::stop(const AttributeMap& extra) {
-        throw std::runtime_error(status.error_message());
-    }
-    return response.is_moving();
+    return make_client_helper(this, *stub_, &StubType::IsMoving).invoke([](auto& response) {
```
Note the lack of a `.with` call here: there are no params and no `extra` so we don't have anything to prepare.

acmorrow commented on src/viam/sdk/common/client_helper.hpp:
```diff
@@ -0,0 +1,82 @@
+
+    template <typename ResponseHandlerCallable, typename ErrorHandlerCallable>
+    auto invoke(ResponseHandlerCallable&& rhc, ErrorHandlerCallable&& ehc) {
+        *request_.mutable_name() = client_->name();
```
Application of `name` is done here so that you are not required to call `with` but can instead directly call `invoke` if you have nothing to configure on the request.

acmorrow commented on src/viam/sdk/components/movement_sensor/client.cpp:
```diff
@@ -26,90 +27,71 @@ MovementSensorClient::MovementSensorClient(std::string name, std::shared_ptr<grp
 
 Vector3 MovementSensorClient::get_linear_velocity(const AttributeMap& extra) {
-    GetLinearVelocityResponse response;
-    stub_wrapper<GetLinearVelocityRequest>(
```
One of the big advantages of `ClientHelper` over `stub_wrapper` is that you only need to know the function on the stub, and you don't need to name the request and response types explicitly (though you can, if you want, on the lambda).

benjirewis commented on src/viam/sdk/common/client_helper.hpp:
```diff
@@ -0,0 +1,82 @@
+    using PFn = ::grpc::Status (StubType::*)(::grpc::ClientContext*,
+                                             const RequestType&,
+                                             ResponseType*);
+    explicit ClientHelper(ClientType* client, StubType* stub, PFn pfn)
```
Why `explicit`?

benjirewis replied:
Sounds good, and thanks for context; just didn't see one on the `ServiceHelper` constructor.

benjirewis commented on src/viam/sdk/common/client_helper.hpp:
```diff
@@ -0,0 +1,82 @@
+    }
+
+   public:
+    using PFn = ::grpc::Status (StubType::*)(::grpc::ClientContext*,
```
[q] I got the other abbreviations, but what does `PFn` stand for?

benjirewis replied:
Ah gotcha; no that seems fine. Could do `Method`, but I don't have strong opinion.

benjirewis commented on src/viam/sdk/common/client_helper.hpp:
```diff
@@ -0,0 +1,82 @@
+    ResponseType response_;
+};
+
+// TODO: Stop returns the status???
```
Did we file a ticket for changing `stop` return types? Happy to do so if one hasn't been made.

acmorrow replied:
@stuqdog ?

stuqdog replied:
yep! https://viam.atlassian.net/browse/RSDK-5760

acmorrow replied:
Thanks! I'll remove the TODO (which shouldn't have moved to here when I pulled this code out to a header anyway).

benjirewis commented on src/viam/sdk/components/movement_sensor/client.cpp:
```diff
@@ -26,90 +27,71 @@ MovementSensorClient::MovementSensorClient(std::string name, std::shared_ptr<grp
-    return Vector3::from_proto(response.linear_velocity());
+    return make_client_helper(this, *stub_, &MovementSensorClient::Stub::GetLinearVelocity)
+        .with(extra)
+        .invoke([](auto& response) { return Vector3::from_proto(response.linear_velocity()); });
```
```suggestion
        .invoke([&](auto& response) { return Vector3::from_proto(response.linear_velocity()); });
```
Why are these `[=]` lambdas when the ones on the motor are `[&]`?

acmorrow replied:
Because the response handler doesn't need access to any captured state, so we don't need `[&]`.  We need it on the `with` lambdas because we need access to the paramers to the function. I don't think motor does it any differently, and I'm pretty sure there aren't any `[=]` lambdas at all.

benjirewis replied:
Ah sounds good; I was seeing `[&]` captures in motor in `with` functions.

> and I'm pretty sure there aren't any [=] lambdas at all.

Oh I thought `[]` implicitly meant `[=]`.

benjirewis commented on src/viam/sdk/common/client_helper.hpp:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    template <typename ResponseHandlerCallable = decltype(default_rhc_)>
+    auto invoke(ResponseHandlerCallable&& rhc = default_rhc_) {
```
Yeah FWIW I appreciate a named `invoke` method over overloading `operator()` because an `invoke` with an rhc and an ehc may be pretty tough to read. The `ServiceHelper` _does_ do the opposite, but it kind of makes sense that the client is the one "invoking" the gRPC endpoint.

benjirewis reviewed (commented):
Awesome; very cool code. Thanks for this, just have some qs!

benjirewis reviewed (approved):
LGTM; thanks for handling these hard pieces!

stuqdog reviewed (approved):
looks great to me, thanks!

acmorrow commented:
@stuqdog - I removed the stale `TODO` comment, added logging for misbehaving `ehc`'s, and went ahead and published the headers for both `ClientHelper` and `ServiceHelper`. I didn't see any reason to hold them back, but I can easily pull that back out if you would like to wait.

stuqdog reviewed (approved):
looks great to me, thanks again!
